### PR TITLE
[TPU] add kv cache update kernel

### PR DESCRIPTION
**Pull request by yaochengji**
## Purpose

TPU is not good at scatter-update. Here consecutive new kv status will be updated together with the help of the kv cache update kernel.

## Test Plan

Kernel test: pytest -s -v tests/v1/tpu/test_kv_cache_update_kernel.py
Accuracy test: pytest -s -v tests/entrypoints/llm/test_accuracy.py::test_lm_eval_accuracy_v1_engine

## Test Result

Passed.

### CODE REVIEW

**Comment by yaochengji** [comment]
The kernel implementation is based on the discussion with @bythew3i  and @vanbasten23 .

We can observe some performance improvement:
Llama 8B: 7.85 reqs/s -> 8.04 reqs/s
Llama 70B: 6.41 reqs/s -> 6.71 reqs/s

File: tests/v1/tpu/test_kv_cache_update_kernel.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    slot_mapping = np.stack(
+        [kv_cache_start_indices, new_kv_cache_indices, slice_lens], axis=1)
+    slot_mapping = np.pad(
+        slot_mapping, [[0, kernel_block_size - slot_mapping.shape[0]], [0, 0]],
+        constant_values=0)
+    slot_mapping_cpu = torch.tensor(slot_mapping, device="cpu")
+    slot_mapping_xla = slot_mapping_cpu.to(torch_xla.device())
+    torch_xla.sync()
+
+    torch.ops.xla.dynamo_set_buffer_donor_(kv_cache_xla, True)
```
Comment (vanbasten23) [inline tests/v1/tpu/test_kv_cache_update_kernel.py:57]: why do we want to do  this `torch.ops.xla.dynamo_set_buffer_donor_`?

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+@functools.partial(
+    jax.jit,
+    static_argnames=["page_size", "block_size"],
+)
+def kv_cache_update(
+    new_kv: jax.Array,  # [total_num_token, kv_head_num, head_dim]
+    slices: jax.
+    Array,  # [num_slices, 3], list of (kv_cache_start, new_kv_start, slice_len)
+    kv_cache: jax.
+    Array,  # [total_num_pages * page_size, kv_head_num, head_dim]
```
Comment (vanbasten23) [inline vllm/attention/ops/pallas_kv_cache_update.py]: nit: use num_combined_kv_heads to be consistent

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        _kv_cache_update_kernel,
+        grid_spec=pltpu.PrefetchScalarGridSpec(
+            num_scalar_prefetch=len(scalar_prefetches),
+            in_specs=in_specs,
+            out_specs=out_specs,
+            grid=(slices.shape[0] // block_size, ),
+            scratch_shapes=scratch_shapes,
+        ),
+        out_shape=out_shape,
+        input_output_aliases={len(scalar_prefetches) + 1: 0},
```
Comment (vanbasten23) [inline vllm/attention/ops/pallas_kv_cache_update.py:114]: I guess this maps kv_cache_hbm_ref to the output so that you don't need to specify the output in "_kv_cache_update_kernel"?

File: vllm/v1/attention/backends/pallas.py [hunk 4]
```diff
@@ -263,4 +272,40 @@ def write_to_kv_cache(
...
+    # NOTE: the in-place copy will be optimized away by XLA compiler.
+    kv_cache.copy_(new_kv_cache)
+
+
+@requires_jax
+def kv_cache_update_op_impl(kv: torch.Tensor, slot_mapping: torch.Tensor,
+                            kv_cache: torch.Tensor, page_size: int,
+                            block_size: int):
+    from vllm.attention.ops.pallas_kv_cache_update import kv_cache_update
+    new_kv_cache = xb.call_jax(kv_cache_update, (kv, slot_mapping, kv_cache), {
```
Comment (vanbasten23) [inline vllm/v1/attention/backends/pallas.py:285]: Very nice!

File: vllm/v1/worker/tpu_model_runner.py [hunk 5]
```diff
@@ -515,6 +513,69 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
...
+                    for the corresponding slice.
+                - new_kv_start_index (int): The starting index in the new KV
+                    cache for the corresponding slice.
+                - slice_len (int): The length of the slice.
+        """
+        slices_start = self.input_batch.num_computed_tokens_cpu[:num_reqs]
+        slices_end = self.input_batch.num_computed_tokens_cpu[:num_reqs] + \
+            num_scheduled_tokens_per_req
+        local_block_start_idx = slices_start // self.block_size
+        local_block_end_idx = (slices_end - 1) // self.block_size
```
Comment (vanbasten23) [inline vllm/v1/worker/tpu_model_runner.py:557]: why is there a "-1" here?

File: tests/v1/tpu/test_kv_cache_update_kernel.py (hunk 1 above)
Comment (yaochengji) [reply tests/v1/tpu/test_kv_cache_update_kernel.py:57]: Because it should be an inplace-update.

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 3 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py:114]: Yeah, they're just aliases.

File: vllm/v1/worker/tpu_model_runner.py (hunk 5 above)
Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:557]: oh nvm, I figured it out.

Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:557]: Because `slices_end` is not included in the range.

File: vllm/v1/worker/tpu_model_runner.py [hunk 6]
```diff
@@ -515,6 +513,69 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
...
+        no_repeat_req_indices = self.arange_np[:num_reqs]
+        global_block_start_idx = (
+            no_repeat_req_indices * self.max_num_blocks_per_req +
+            local_block_start_idx)
+        block_lens = local_block_end_idx - local_block_start_idx + 1
+        global_block_start_idx = np.repeat(global_block_start_idx, block_lens)
+        slice_arange = np.concatenate([self.arange_np[:n] for n in block_lens])
+        global_block_indices = global_block_start_idx + slice_arange
+        block_table_cpu = self.input_batch.block_table[0].get_cpu_tensor()
```
Comment (vanbasten23) [inline vllm/v1/worker/tpu_model_runner.py:566]: do you know why we only get the first element of input_batch.block_table?

File: tests/v1/tpu/test_kv_cache_update_kernel.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    new_kv_cache_xla = torch.ops.xla.kv_cache_update_op(
+        new_kv_xla, slot_mapping_xla, kv_cache_xla, page_size,
+        kernel_block_size)
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply tests/v1/tpu/test_kv_cache_update_kernel.py]: The explanation is already in the pallas kernel.

File: vllm/v1/worker/tpu_model_runner.py [hunk 8]
```diff
@@ -623,6 +658,18 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
             self.device)
         seq_lens = self.seq_lens_cpu[:self.max_num_reqs].to(self.device)
 
+        slot_mapping_metadata = self._get_slot_mapping_metadata(
+            num_reqs, num_scheduled_tokens_per_req)
+        padded_num_slices = _get_padded_num_kv_cache_update_slices(
+            padded_total_num_scheduled_tokens, self.max_num_reqs,
+            self.block_size)
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:718]: It's obvious, comments are not needed.

File: vllm/v1/worker/tpu_model_runner.py [hunk 9]
```diff
@@ -1646,6 +1697,19 @@ def _get_padded_token_len(paddings: list[int], x: int) -> int:
...
+                                           page_size: int) -> int:
+    """ A fixed shape of slot_mapping_metadata tensor is required to avoid
+    recompilation.
+    """
+    padded_num_slices = 2 * max_num_reqs + num_tokens // page_size
+    pagged_num_slices = min(padded_num_slices, num_tokens)
+    pagged_num_slices = (
+        pagged_num_slices + KV_CACHE_UPDATE_BLOCK_SIZE -
+        1) // KV_CACHE_UPDATE_BLOCK_SIZE * KV_CACHE_UPDATE_BLOCK_SIZE
+    return pagged_num_slices
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py]: comments are not needed.

File: vllm/v1/worker/tpu_model_runner.py (hunk 6 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:566]: I remember it's a list of list for more complex uses, but I don't know too much details. Here the logic is basically reusing the old one.

File: tests/v1/tpu/test_kv_cache_update_kernel.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+@pytest.mark.skipif(not current_platform.is_tpu(),
+                    reason="This is a test for TPU only")
+def test_kv_cache_update_kernel():
+    page_num = 1000
+    page_size = 32
+    combined_kv_head_num = 16
+    head_dim = 128
+    kernel_block_size = 16
+    padded_num_tokens = 128
```
Comment (mgoin) [inline tests/v1/tpu/test_kv_cache_update_kernel.py]: Are there reasonable edge cases we could test here? Like block size != 16, an odd number of tokens, small kv head num, etc

File: vllm/v1/worker/tpu_model_runner.py [hunk 11]
```diff
@@ -53,12 +53,11 @@
...
 logger = init_logger(__name__)
 
-# Here we utilize the behavior that out-of-bound index is ignored.
-# FIXME(woosuk): Find a more reliable way to prevent possible bugs.
-_PAD_SLOT_ID = 1_000_000_000
 INVALID_TOKEN_ID = -1
 # Smallest output size
 MIN_NUM_SEQS = 8
+# Block size used for kv cache updating kernel
+KV_CACHE_UPDATE_BLOCK_SIZE = 8
```
Comment (mgoin) [inline vllm/v1/worker/tpu_model_runner.py]: It would be good to specify what dim this is blocking. For instance, why isn't it the same as the kv cache block size? What is the reason for not tuning this higher?

File: vllm/v1/worker/tpu_model_runner.py [hunk 12]
```diff
@@ -1646,6 +1698,18 @@ def _get_padded_token_len(paddings: list[int], x: int) -> int:
...
+def _get_padded_num_kv_cache_update_slices(num_tokens: int, max_num_reqs: int,
+                                           page_size: int) -> int:
+    """Calculates the padded number of KV cache update slices to avoid
+    recompilation."""
+    padded_num_slices = 2 * max_num_reqs + num_tokens // page_size
+    pagged_num_slices = min(padded_num_slices, num_tokens)
+    pagged_num_slices = (
+        pagged_num_slices + KV_CACHE_UPDATE_BLOCK_SIZE -
+        1) // KV_CACHE_UPDATE_BLOCK_SIZE * KV_CACHE_UPDATE_BLOCK_SIZE
+    return pagged_num_slices
```
Comment (mgoin) [inline vllm/v1/worker/tpu_model_runner.py]:
```suggestion
    padded_num_slices = min(padded_num_slices, num_tokens)
    padded_num_slices = (
        padded_num_slices + KV_CACHE_UPDATE_BLOCK_SIZE -
        1) // KV_CACHE_UPDATE_BLOCK_SIZE * KV_CACHE_UPDATE_BLOCK_SIZE
    return padded_num_slices
```

**Review by mgoin — COMMENTED** [review]
LGTM if the new jax imports don't disturb the other backends in CI. Just a few commnets

File: vllm/v1/worker/tpu_model_runner.py [hunk 13]
```diff
@@ -1646,6 +1697,19 @@ def _get_padded_token_len(paddings: list[int], x: int) -> int:
     return paddings[index]
 
 
+def _get_padded_num_kv_cache_update_slices(num_tokens: int, max_num_reqs: int,
+                                           page_size: int) -> int:
+    """ A fixed shape of slot_mapping_metadata tensor is required to avoid
+    recompilation.
+    """
+    padded_num_slices = 2 * max_num_reqs + num_tokens // page_size
```
Comment (vanbasten23) [inline vllm/v1/worker/tpu_model_runner.py:1804]: could you explain these 2 lines
```
    padded_num_slices = 2 * max_num_reqs + num_tokens // page_size
    pagged_num_slices = min(padded_num_slices, num_tokens)
```
Why is pagged_num_slices determined this way?

File: tests/v1/tpu/test_kv_cache_update_kernel.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@pytest.mark.skipif(not current_platform.is_tpu(),
+                    reason="This is a test for TPU only")
+def test_kv_cache_update_kernel():
+    page_num = 1000
+    page_size = 32
+    combined_kv_head_num = 16
+    head_dim = 128
+    kernel_block_size = 16
```
Comment (vanbasten23) [inline tests/v1/tpu/test_kv_cache_update_kernel.py]: could you also test when the kernel need to iterate thru multiple blocks?

File: vllm/v1/worker/tpu_model_runner.py [hunk 15]
```diff
@@ -515,6 +513,69 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
...
+        Returns:
+            np.ndarray: A 2D array of shape (total_block_len, 3), where each row
+            contains:
+                - kv_cache_start_index (int): The starting index in the KV cache
+                    for the corresponding slice.
+                - new_kv_start_index (int): The starting index in the new KV
+                    cache for the corresponding slice.
+                - slice_len (int): The length of the slice.
+        """
+        slices_start = self.input_batch.num_computed_tokens_cpu[:num_reqs]
```
Comment (vanbasten23) [inline vllm/v1/worker/tpu_model_runner.py:553]: I wonder which test did you run this test this change.

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+import functools
+
+import jax
+from jax.experimental import pallas as pl
+from jax.experimental.pallas import tpu as pltpu
+
+
+def _kv_cache_update_kernel(
+    # Prefetch
+    slices_ref,  # [num_slices, 3]
```
Comment (bythew3i) [inline vllm/attention/ops/pallas_kv_cache_update.py]: Consider making it to [3, num_slices] or [3 * num_slices] to save smem

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    new_kv: jax.Array,  # [total_num_token, num_combined_kv_heads, head_dim]
+    slices: jax.
+    Array,  # [num_slices, 3], list of (kv_cache_start, new_kv_start, slice_len)
+    kv_cache: jax.
+    Array,  # [total_num_pages * page_size, num_combined_kv_heads, head_dim]
+    *,
+    page_size: int = 32,
+    block_size: int = 8,
+):
+    assert slices.shape[0] % block_size == 0
```
Comment (vanbasten23) [inline vllm/attention/ops/pallas_kv_cache_update.py]: do you need to check other precondition such as new_kv.shape[1:]==kv_cache.shape[1:]?
Also, is there any constraint for the slice_len?

File: tests/v1/tpu/test_kv_cache_update_kernel.py (hunk 10 above)
Comment (yaochengji) [reply tests/v1/tpu/test_kv_cache_update_kernel.py]: Thanks Michael for the great suggestion. I added different head_dim, page_size, and kv_head_num. The decoding, prefill within a page, prefill across pages cases are already in the test.

File: vllm/v1/worker/tpu_model_runner.py (hunk 11 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py]: It's the kernel block size, which is different from the block size in vLLM (A.K.A page size). I set it to 8 because the min_num_seqs is 8, and I'd like to avoid trivial DMA is possible.

File: vllm/v1/worker/tpu_model_runner.py (hunk 12 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py]: Done. Thanks for catching the typo.

File: vllm/v1/worker/tpu_model_runner.py (hunk 13 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:1804]: Because the max possible number of slices is limited by num_reqs and num_tokens.

File: tests/v1/tpu/test_kv_cache_update_kernel.py (hunk 14 above)
Comment (yaochengji) [reply tests/v1/tpu/test_kv_cache_update_kernel.py]: Thanks for the great suggestion! It's added.

File: vllm/v1/worker/tpu_model_runner.py (hunk 15 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:553]: In the e2e accuracy test.

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 17 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py]: Yeah, I added other necessary  cheks.

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+def _kv_cache_update_kernel(
+    # Prefetch
+    slices_ref,  # [num_slices, 3]
+    # Input
+    new_kv_hbm_ref,  # [tokens, num_combined_kv_heads, head_dim]
+    kv_cache_hbm_ref,
+    # Output
+    _,  # [total_num_pages * page_size, num_combined_kv_heads, head_dim]
+    # Scratch
+    scratch,  # [block_size, page_size, num_combined_kv_heads, head_dim]
```
Comment (bythew3i) [inline vllm/attention/ops/pallas_kv_cache_update.py]: Please rename the `block_size` to something more meaningful like "kv_pages_per_blk"? Because block_size is overused, eg., it means page_size in vllm which will be very confused for readers

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+import jax
+from jax.experimental import pallas as pl
+from jax.experimental.pallas import tpu as pltpu
+
+
+def _kv_cache_update_kernel(
+    # Prefetch
+    slices_ref,  # [num_slices, 3]
+    # Input
+    new_kv_hbm_ref,  # [tokens, num_combined_kv_heads, head_dim]
```
Comment (bythew3i) [inline vllm/attention/ops/pallas_kv_cache_update.py]: Nit: `tokens` to `num_tokens`?

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+    kv_cache_hbm_ref,
```
Comment (bythew3i) [inline vllm/attention/ops/pallas_kv_cache_update.py]: Nit: Add kv_cache shape in comment

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+    block_size = scratch.shape[0]
+
+    # Copy from new_kv_hbm_ref to scratch
+    for i in range(block_size):
+        offset_i = i + block_idx * block_size
+        new_kv_start = slices_ref[offset_i, 1]
+        length = slices_ref[offset_i, 2]
+        async_copy = pltpu.make_async_copy(
+            new_kv_hbm_ref.at[pl.ds(new_kv_start, length), ...],
+            scratch.at[i, pl.ds(0, length), ...],
```
Comment (bythew3i) [inline vllm/attention/ops/pallas_kv_cache_update.py:37]: Where do we guarantee that the length is not larger than the page_size?

File: vllm/attention/ops/pallas_kv_cache_update.py [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+        )
+        async_copy.start()
+        async_copies.append(async_copy)
+
+    for async_copy in async_copies:
+        async_copy.wait()
+
+    # Copy from scratch to kv_cache_hbm_ref
+    async_copies.clear()
+    for i in range(block_size):
```
Comment (bythew3i) [inline vllm/attention/ops/pallas_kv_cache_update.py:45]: Why did we just iterate through `num_slices` to update?

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 18 above)
Comment (bythew3i) [reply vllm/attention/ops/pallas_kv_cache_update.py]: By reading the code, I think it might be better to rename `block_size` to `num_slices_per_blk`

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 16 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py]: Done. So there's some kind of padding for smem?

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 18 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py]: Done. Thanks for the awesome suggestion!

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 19 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py]: Done.

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 20 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py]: Done.

File: vllm/attention/ops/pallas_kv_cache_update.py (hunk 21 above)
Comment (yaochengji) [reply vllm/attention/ops/pallas_kv_cache_update.py:37]: Users of the kernel should guarantee the input is valid.
I added a TODO for the dynamic input check.

File: vllm/v1/worker/tpu_model_runner.py (hunk 11 above)
Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py]: or would it be better to put "kernel" in the name, such as KV_CACHE_KERNEL_BLOCK_SIZE to distinguish it from the block size in vLLM (page size)?

Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py]: I changed it to `NUM_SLICES_PER_KV_CACHE_UPDATE_BLOCK` already.

File: vllm/v1/worker/tpu_model_runner.py (hunk 13 above)
Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:1804]: Not sure if I follow. I understand the number of slice can be bounded by `num_tokens`. Could we just do  `pagged_num_slices = num_tokens`?

I'm not sure where `2 * max_num_reqs + num_tokens // page_size` comes from

Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:1804]: Let's say there're max_num_reqs requests, and each request has `R_i` tokens. Also each request has at most `2 + R_i // page_size` pages. The total tokens for all the requests are `num_tokens`.

So the page_num = sum(2 + R_i // page_size) <= 2 * max_num_reqs + sum(R_i) // page_size = 2 * max_num_reqs + num_tokens // page_size

Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:1804]: Regarding why `each request has at most 2 + R_i // page_size pages`, we chatted offline. Here we are estimating a upper bound of num_pages for a request. If there are 2 tokens and these 2 tokens are at the page boundary, then it needs 2 pages.
